Gems/GraphModel/Code/Source/Model/Slot.cpp

```diff
@@ -601,71 +606,65 @@ namespace GraphModel
+        // Because some slots support multiple data types search for the one corresponding to the value
+        return GetDataTypeForValue(m_value);
     }
 
     void Slot::SetValue(const AZStd::any& value)
     {
         if (SupportsValue())
         {
 #if defined(AZ_ENABLE_TRACING)
-            DataTypePtr dataType = GetGraphContext()->GetDataTypeForValue(value);
-            AssertTypeMatch(dataType, "Slot::SetValue used with the wrong type");
+            DataTypePtr dataTypeUsed = GetDataTypeForValue(value);
+            AssertWithTypeInfo(SupportsValue(), dataTypeUsed, "This slot type does not support values");
+            AssertWithTypeInfo(IsSupportedDataType(dataTypeUsed), dataTypeUsed, "Slot::SetValue used with the wrong type");
 #endif
-
             m_value = value;
         }
     }
 
 #if defined(AZ_ENABLE_TRACING)
     void Slot::AssertWithTypeInfo(bool expression, DataTypePtr dataTypeUsed, const char* message) const
     {
+        DataTypePtr dataType = GetDataType();
         AZ_Assert(expression, "%s (Slot DataType=['%s', '%s', %s]. Used DataType=['%s', '%s', %s]). m_value TypeId=%s.",
             message,
-            GetDataType()->GetDisplayName().c_str(),
-            GetDataType()->GetCppName().c_str(),
-            GetDataType()->GetTypeUuidString().c_str(),
-            dataTypeUsed->GetDisplayName().c_str(),
-            dataTypeUsed->GetCppName().c_str(),
-            dataTypeUsed->GetTypeUuidString().c_str(),
+            !dataType ? "nullptr" : dataType->GetDisplayName().c_str(),
+            !dataType ? "nullptr" : dataType->GetCppName().c_str(),
+            !dataType ? "nullptr" : dataType->GetTypeUuidString().c_str(),
+            !dataTypeUsed ? "nullptr" : dataTypeUsed->GetDisplayName().c_str(),
+            !dataTypeUsed ? "nullptr" : dataTypeUsed->GetCppName().c_str(),
+            !dataTypeUsed ? "nullptr" : dataTypeUsed->GetTypeUuidString().c_str(),
             m_value.type().ToString<AZStd::string>().c_str()
         );
     }
+#endif // AZ_ENABLE_TRACING
 
-    void Slot::AssertTypeMatch(DataTypePtr dataTypeUsed, const char* message) const
+    DataTypePtr Slot::GetDataTypeForTypeId(const AZ::Uuid& typeId) const
     {
-        // Check if any of the possible data types for this slot match
-        bool expression = false;
-        for (auto iter : GetPossibleDataTypes())
+        for (DataTypePtr dataType : GetSupportedDataTypes())
         {
-            expression |= (*dataTypeUsed == *iter);
+            if (!SupportsDataType() || dataType->IsSupportedType(typeId))
+            {
+                return dataType;
+            }
         }
-        AssertWithTypeInfo(expression, dataTypeUsed, message);
+
+        return {};
     }
-#endif // AZ_ENABLE_TRACING
 
+    DataTypePtr Slot::GetDataTypeForValue(const AZStd::any& value) const
+    {
+        for (DataTypePtr dataType : GetSupportedDataTypes())
+        {
+            if (!SupportsValue() || dataType->IsSupportedValue(value))
```

Comment:
I would have expected these functions to return "{}" when the slot doesn't support values. Maybe this deserves a comment to explain why.

Reply:
https://github.com/o3de/o3de/pull/10390